source/ceylon/future/Future.ceylon

```diff
@@ -0,0 +1,92 @@
+import ceylon.future.util {
+    Latch,
+    AcquireTimeoutException
+}
+import ceylon.promise {
+    Promise,
+    ExecutionContext
+}
+
+"Wraps a [[Promise]] that will be completed by another thread."
+see (`function submit`)
+shared class Future<Value>(delegate)
+        extends Promise<Value>() {
+    "The concurrently completed [[Promise]]."
+    Promise<Value> delegate;
+    
+    /*
+     * Delegation
+     */
+    
+    shared actual ExecutionContext context => delegate.context;
+    
+    shared actual Promise<Result> map<Result>(
+        Result onFulfilled(Value val),
+        Result onRejected(Throwable reason))
+            => delegate.map(onFulfilled, onRejected);
+    
+    shared actual Promise<Result> flatMap<Result>(
+        Promise<Result> onFulfilled(Value val),
+        Promise<Result> onRejected(Throwable reason))
+            => delegate.flatMap(onFulfilled, onRejected);
+    
+    /*
+     * New functionality
+     */
+    
+    value completion = Latch();
+    
+    variable Boolean _done = false;
+    "[[true]] if the [[delegate]] is completed."
+    shared Boolean done => _done;
+    
+    variable Value? resultValue = null;
+    variable Throwable? resultException = null;
```

Comment:
Are `resultValue` and `resultException` not already mutexed by the `completion` Latch? They're not shared and are written to at most once before `completion` releases, and only read after `completion` releases.